docs/proposals/20230612-node-os-image-caching.md

```diff
@@ -164,53 +165,82 @@ metadata:
 spec:
   template:
     nodeCachingMode:
-      interval: 1 day
+      interval: 24h
 ```
 
-A model scenario would be an operator spinning up a CAPZ cluster and having this feature be able to be toggled on or off. If it was toggled on, then as the months passed and more security updates and patches needed to be applied to the operator’s node OS image, these changes would be cached on a regular interval and the operator would no longer have to wait for these changes to apply on new node creations. This would help prevent warm nodes and overprovisioning as well as allowing the operator to have faster horizontal scaling.
-
 ![Figure 1](./images/node-os-image-cache.png)
 
 ### Security Model
-This proposal requires CAPZ to have write permissions for azureMachineTemplates in order to properly update the nodes’ OS image on the spec. Regex expressions will potentially be used for when a user declares a time amount for the interval that they want to specify, and denial of service attacks will be protected against by having an update system which doesn’t need to be atomic, if part of the caching is complete there is no risk in the update not finishing since the spec update will happen at once. No sensitive data is being stored in a secret. 
+This proposal requires CAPZ to have write permissions for azureMachineTemplates in order to properly update the nodes’ OS image on the spec. Go has a library called time and time.ParseDuration will be used to parse the time interval provided by an operator instead of using regular expressions. Denial of service attacks will be protected against by having an update system which doesn’t need to be atomic. If part of the caching is complete there is no risk in the update not finishing since the spec update will happen at once. No sensitive data is being stored in a secret.
 
 ### Risks and Mitigations
 
 Example risks:
-1. A bad snapshot is taken, and we will mitigate this risk by having a rollback if a bad snapshot is detected as taken and have prevention techniques in place to detect if the snapshot is bad before taking it.
+1. A bad snapshot is taken, and we will mitigate this risk by having a rollback if a bad snapshot is detected as taken and have prevention techniques in place to detect if the snapshot is bad before taking it. In practice, when a new node is spun up with the bad snapshot, it will be detected and then the node will be shut down. The node OS image being used to create new nodes will then be reverted to the previous one which will still be cached until a successful new node has been spawned, so since it was a failed creation here we will just revert to that previous image and create the desired node.
```

Comment:
> when a new node is spun up with the bad snapshot, it will be detected

Detecting which nodes are "bad" might be a bit gnarly for CAPZ to do generically. Maybe we can call this out of scope and put this responsibility on the user? So a user might ensure a small scale out with a new image is successful (by their own metrics) before scaling further out. Or perhaps this might be integrated with MachineHealthChecks (which I think might also be a bit too ambitious for now)?

Reply:
Moved it to out of scope and noted that we will simply try to avoid it as best as we currently can and the rest is on the operator